src/reference_implementations/prompting_vector_llms/prompt_ensembling/copa_prompting_examples.ipynb

```diff
@@ -0,0 +1,1036 @@
+    "# Beginning of sentence token (\"BOS\")\n",
+    "encoded_tokens = tokenizer.encode(\"Hello this is a test\")\n",
+    "print(f\"Encoded Tokens: {encoded_tokens}\")\n",
+    "# If you ever need to move back from token ids, you can use tokenizer.decode or tokenizer.batch_decode\n",
+    "decoded_tokens = tokenizer.decode(encoded_tokens)\n",
+    "print(f\"Decoded Tokens: {decoded_tokens}\")\n",
+    "\n",
+    "# We're interested in the activations from the last layer of the model, because this will allow us to calculate the\n",
+    "# likelihoods\n",
+    "last_layer_name = llama_model.module_names[-1]\n",
+    "print(f\"Last Layer Name: {last_layer_name}\")\n",
+    "# Get a log softmax function to compute log probabilities from the output layer.\n",
+    "log_softmax = nn.LogSoftmax(dim=1)\n",
+    "\n",
+    "endline_token_id = tokenizer.encode(\"Hello\\n\")[-1]\n",
+    "print(f\"Endline Token Id: {endline_token_id}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def compute_log_probability_from_activations(logits: torch.Tensor, token_ids: List[int]) -> float:\n",
+    "    # First we get the logprobs associated with each token, logits is n_tokens x vocabulary size\n",
+    "    log_probs = log_softmax(logits.type(torch.float32))\n",
+    "    # Drop the first token ID (as it corresponds to the <s> token) and add placeholder to the end. This shift aligns\n",
+    "    # The tokens with the output activations corresponding to their logprobs\n",
+    "    token_ids.pop(0)\n",
+    "    token_ids.append(1)\n",
+    "    # We only really care about the logprobs associated with the sentence to be completed\n",
+    "    # (i.e. not the demonstrations or the question). So search for the last endline in the tokens and only\n",
+    "    # sum the logprobs thereafter.\n",
+    "    endline_index = len(token_ids) - list(reversed(token_ids)).index(endline_token_id)\n",
+    "    # Turn token ids into the appropriate column indices\n",
+    "    token_id_slicer = torch.Tensor(token_ids).reshape(-1, 1).type(torch.int64)\n",
+    "    log_probs_per_token = log_probs.gather(1, token_id_slicer)\n",
+    "    # We sum the log probabilities, except for the last one which corresponds to the as yet predicted token)\n",
+    "    # and then normalize by the number of tokens (minus one for the placeholder)\n",
+    "    selected_log_probs_per_token = log_probs_per_token[endline_index:-1]\n",
+    "    normalized_log_prob = torch.sum(selected_log_probs_per_token) / len(selected_log_probs_per_token)\n",
+    "    return normalized_log_prob.item()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 28,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We're running a lot of activation retrievals. Once in a while there is a json decoding or Triton error. If that\n",
+    "# happens, we retry the activations request.\n",
```

Comment:
I assume these errors are due to the underlying Klaidescope implementation and it is normal to simply retry?

Reply:
Yeah, the error is a bit ethereal in the sense that if you run the same query again, it doesn't fail. It likely has something to do with the Triton server on the backend. So we're just sort of ignoring it as a fundamental issue for now.